base/src/commonMain/kotlin/state/State.kt

```diff
@@ -17,64 +17,201 @@ Copyright 2019 Splendo Consulting B.V. The Netherlands
 
 */
 
-import com.splendo.kaluga.flow.BaseFlowable
+import com.splendo.kaluga.base.flow.ColdFlowable
+import com.splendo.kaluga.base.flow.HotFlowable
 import com.splendo.kaluga.base.runBlocking
+import com.splendo.kaluga.flow.BaseFlowable
+import com.splendo.kaluga.flow.FlowConfig
 import kotlinx.coroutines.*
+import kotlinx.coroutines.flow.Flow
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * State to be represented in a state machine
+ *
+ * @param T type of the State
+ * @param repoAccessor The [StateRepoAccesor] for accessing the [StateRepo] associated with this State
+ */
+open class State<T:State<T>>(open val repoAccessor:StateRepoAccesor<T>){
 
-open class State<T:State<T>>(open val repo:StateRepo<T>){
+    /**
+     * Called when this state is the first state of the state machine
+     */
+    open suspend fun initialState() {}
+
+    /**
+     * Called while transitioning to a new state before the new state is created
+     */
     open suspend fun beforeCreatingNewState() {}
-    open suspend fun afterCreatingNewState() {}
+    /**
+     * Called while transitioning to a new state after the new state is created
+     *
+     * @param newState the newly created state
+     */
+    open suspend fun afterCreatingNewState(newState: T) {}
+
+    /**
+     * Called while transitioning to a new state after the new state is set.
+     */
     open suspend fun afterNewStateIsSet() {}
+
+    /**
+     * Called while transitioning from an old state before it is removed.
+     */
+    open suspend fun beforeOldStateIsRemoved() {}
+
+    /**
+     * Called while transitioning from an old state after it is removed
+     *
+     * @param oldState the removed state
+     */
+    open suspend fun afterOldStateIsRemoved(oldState: T) {}
+
+    /**
+     * Called when this state is the final state of the state machine
+     */
+    open suspend fun finalState() {}
+
+    /**
+     * Changes from this state to a new state. Only transitions if this State is the current state of the state machine.
+     *
+     * @param toState the state to which to transition
+     */
+    protected suspend fun changeState(toState: T) {
+        repoAccessor.changeState {
+            if (it === this)
+                toState
+            else
+                it
```

Comment:
I'd like to have `{` and `}` for single lines in if/else, now this `it` looks so alone 😢

Reply:
Kotlin is very biased towards single line bracketless statements (e.g. think of `when` mappings, `=` based method declarations, etc. Should we really not be able to do this?